packages/next-s3-upload/src/hooks/use-s3-upload.tsx

```diff
@@ -78,36 +81,41 @@ export const useS3Upload = () => {
       //   queueSize: 1,
       // };
 
-      let s3Upload = s3.upload(params);
+      let s3Upload = new Upload({
+        client,
+        params
+      });
 
       setFiles(files => [
         ...files,
         { file, progress: 0, uploaded: 0, size: file.size },
       ]);
 
-      s3Upload.on('httpUploadProgress', event => {
+      s3Upload.on('httpUploadProgress', (event: any) => {
         if (event.total) {
           setFiles(files =>
             files.map(trackedFile =>
               trackedFile.file === file
                 ? {
-                    file,
-                    uploaded: event.loaded,
-                    size: event.total,
-                    progress: (event.loaded / event.total) * 100,
-                  }
+                  file,
+                  uploaded: event.loaded,
+                  size: event.total,
+                  progress: (event.loaded / event.total) * 100,
+                }
                 : trackedFile
             )
           );
         }
       });
 
-      let uploadResult = await s3Upload.promise();
+      await s3Upload.done();
+
+      const location = `https://s3.amazonaws.com/${data.bucket}/${data.key}`;
```

Comment:
- The only thing that I don't like so much is that I had to generate the URL myself, although the format seems simple enough - v3 of the sdk doesn't seem to provide a "Location" property that existed in v2 of the sdk.

Reply:
I think it's best to find a way to have AWS give us the url, since this seems like knowledge next-s3-upload shouldn't be aware of.

1) What does `s3Upload.done()` return? Anything useful?
2) If we can't get the URL from `s3Upload` is there anyway to get it from AWS's S3 lib given that we know both the `bucket` and `key`?

Reply:
OK, this turns out to be a regression in v3 with a fix posted last August: https://github.com/aws/aws-sdk-js-v3/pull/2700. Some other related open issues: https://github.com/aws/aws-sdk-js-v3/issues/2694 and https://github.com/aws/aws-sdk-js-v3/discussions/2675.
Long story short, I was able to get this to work by down-grading @aws-sdk/lib-storage to v 0.25.0. I tried finding an alternative method to retrieve url but couldn't find anything.
I'll update my PR with the version that works.

Reply:
Oh, I think `@aws-sdk/lib-storage` is mostly a wrapper around the main SDK. You could also rewrite the s3 upload using `@aws-sdk/s3-client` directly, in theory. I tried, but it was hard to find examples.

Reply:
One more thing, to make it work in Next.js, to get the older version to work, I had to add this to my `next.config.js` or else it would get an error on compile: "cannot find "fs" module":
```
  webpack (config, { isServer }) {
    // fix for @aws-sdk/storage-lib v.26 until this issue is fixed:
    if (!isServer) {
      _config.resolve.fallback.fs = false;
    }
    return config;
  }
```
The newer version of `lib-storage` didn't require this step, so there's not really a best all-around solution right now.

Reply:
Ok cool, thanks for digging into this!

The webpack issue is because there's something trying to `import 'fs'` in the browser bundle, and that module doesn't exist in browsers so it blows up.

I think asking people to modify their webpack config in order to use this addon is a lot. There may be some ways we can automatically modify webpack, but right now I'm not sure. The reason I want to avoid asking people to do is because this is the kind of thing that leads to maintenance nightmares in the future.

Other than that I see no problem using the older version of lib-storage.


Ideally I'd like to see AWS fix the issue, but it looks like that PR has gone stale :(

Reply:
One thing about pinning to a specific version would mean that we would force the users for this library to always bundle the version.

This means if someone needs a version of lib-storage (or aws-sdk) higher than 3.25.0 then they might not be able to install it. That, or they end up with multiple copies of aws-sdk in their build (which could create other problems).

I asked around about this and the answer I got was that we really should support all of aws-sdk v3 and not only a specific older version. If the newer versions of v3 are broken then we can't say this library works with v3 without creating a bunch of issues.

I'm kind of convinced this needs to be fixed on the AWS side for us to officially support v3. I'll keep an eye on the issues you linked and see if they say anything. The other option would be use to the S3Client (and not lib-storage), but that looks unfortunately complicated.

Sorry I know that's not the answer you wanted to hear.

Reply:
> The other option would be use to the S3Client (and not lib-storage), but that looks unfortunately complicated.
I agree with this as the best way forward, for anyone reading this that wants to contribute.
Anyway, the file size issue is worth the trade-off to me so I ended up just copy/pasting what I needed, but thanks again because this lib and the AWS setup tutorial saved me a lot of time getting started.

Reply:
Ok cool. Let's leave this PR open as I think it is the right approach and a good reference for anyone wanting to get their app on v3.

I'll check in on that lib-storage bugfix every few weeks and see if it gets resolved.